src/vmm/src/devices/virtio/queue.rs

```diff
@@ -231,6 +276,9 @@ pub struct Queue {
     pub num_added: Wrapping<u16>,
 }
 
+/// SAFETY: Only one thread will access this at a time
```

Comment:
1. Queues are only accessed by vmm thread. You might be thinking about mmio region that can be accessed by vcpus and vmm threads.
2. Pointers are not `Send` by default.

Reply:
`Send` just means we can send instance of the type to another thread (and we do it with `Arc<Mutex<Device..>>` thing inside the `Bus`). But if we "send" the object, it means that we should not have write access to it (until type is also `Sync`). Before we did not need `Send` because all types were Rust verifiable (with lifetimes). But pointers are not. By saying that we only access queue in one thread, we say that sending these pointers is fine, as no one will use them except vmm thread.
At least this is my understanding. It is hard for my C++ brain to understand all intricacies of Rust memory model.

Reply:
> By saying that we only access queue in one thread, we say that sending these pointers is fine, as no one will use them except vmm thread.

If I understand the semantics correctly, by saying we only access the type from one thread we say: "We don't care if this is `Send` because we won't "Send" it anyways, but we mark it `Send` so that the compiler stops complaining. If that is the case, it sounds wrong. Because, the moment the assumption (access from one thread only) changes (e.g. because someone comes and adds multi-threading IO handling) this will be fundamentally broken.

Reply:
Ok, what is your suggestion for this then?

Reply:
I suggest you understand if the code here is `Send` and if it is document why. If not, mention clearly in the comment why it is not.

FWIW, I think the type is indeed `Send` (and `Sync`).
`Send` because this `Queue` is the only object holding these pointers, and `Sync` because it doesn't allow for interior mutability.

Reply:
I think the fact that only VMM thread will access `Queue`s is enough to mark it `Send`. Even thought we actually do not "Send" queues as they are inside devices which are allocated on the heap. The only thing we "Send" are the `Arc<Mutex<....>>`.

Reply:
@roypat as an Rust safety and correctness ambassador, do you have a better resolution for this issue?

Reply:
shouldnt the same logic as with the iovec stuff apply here?

Reply:
`IoVecBuffer` does not allow for interior mutability. `Queue` does. (not really `interior`, but close to it)

Reply:
So, we can hold a `&Queue` reference and change the pointer? Also, IIUC the interior mutability thing is a requirement for `Sync` not `Send`. My understanding coming from the nomicon: https://doc.rust-lang.org/nomicon/send-and-sync.html

> Something can safely be Send unless it shares mutable state with something else without enforcing exclusive access to it. Each Carton has a unique pointer, so we're good.

and

> What about Sync? For Carton to be Sync we have to enforce that you can't write to something stored in a &Carton while that same something could be read or written to from another &Carton. Since you need an &mut Carton to write to the pointer

Reply:
> So, we can hold a &Queue reference and change the pointer?

Reply:
But because all methods that use pointers and mutate memory need `&mut Queue` maybe everything is fine.

Reply:
> IoVecBuffer does not allow for interior mutability.
IoVecBufferMut then

Reply:
Settled on new safety comment.